62_CAD/shaders/main_pipeline/fragment_shader.hlsl

```diff
@@ -1,655 +1,1104 @@
+
+    const uint32_t packedData = pseudoStencil[fragCoord];
+
+    const uint32_t localQuantizedAlpha = (uint32_t)(localAlpha * 255.f);
+    const uint32_t storedQuantizedAlpha = nbl::hlsl::glsl::bitfieldExtract<uint32_t>(packedData,0,AlphaBits);
+    const uint32_t storedMainObjectIdx = nbl::hlsl::glsl::bitfieldExtract<uint32_t>(packedData,AlphaBits,MainObjectIdxBits);
+    // if geomID has changed, we resolve the SDF alpha (draw using blend), else accumulate
+    const bool differentMainObject = currentMainObjectIdx != storedMainObjectIdx; // meaning current pixel's main object is different than what is already stored
+    const bool resolve = differentMainObject && storedMainObjectIdx != InvalidMainObjectIdx;
+    uint32_t toResolveStyleIdx = InvalidStyleIdx;
+    
+    // load from colorStorage only if we want to resolve color from texture instead of style
+    // sampling from colorStorage needs to happen in critical section because another fragment may also want to store into it at the same time + need to happen before store
+    if (resolve)
+    {
+        toResolveStyleIdx = loadMainObject(storedMainObjectIdx).styleIdx;
+        if (toResolveStyleIdx == InvalidStyleIdx) // if style idx to resolve is invalid, then it means we should resolve from color
+            color = float32_t4(unpackR11G11B10_UNORM(colorStorage[fragCoord]), 1.0f);
+    }
+    
+    // If current localAlpha is higher than what is already stored in pseudoStencil we will update the value in pseudoStencil or the color in colorStorage, this is equivalent to programmable blending MAX operation.
+    // OR If previous pixel has a different ID than current's  (i.e. previous either empty/invalid or a differnet mainObject), we should update our alpha and color storages.
+    if (differentMainObject || localQuantizedAlpha > storedQuantizedAlpha)
+    {
+        pseudoStencil[fragCoord] = nbl::hlsl::glsl::bitfieldInsert<uint32_t>(localQuantizedAlpha,currentMainObjectIdx,AlphaBits,MainObjectIdxBits);
+        if (colorFromTexture) // writing color from texture
+            colorStorage[fragCoord] = packR11G11B10_UNORM(localTextureColor);
+    }
+    
+    nbl::hlsl::spirv::endInvocationInterlockEXT();
+
+    if (!resolve)
+        discard;
+
+    // draw with previous geometry's style's color or stored in texture buffer :kek:
+    // we don't need to load the style's color in critical section because we've already retrieved the style index from the stored main obj
+    if (toResolveStyleIdx != InvalidStyleIdx) // if toResolveStyleIdx is valid then that means our resolved color should come from line style
+        color = loadLineStyle(toResolveStyleIdx).color;
+    color.a *= float(storedQuantizedAlpha) / 255.f;
+    
+    return color;
+}
+
+float dot2(in float2 vec)
+{
+    return dot(vec, vec);
+}
+
+
+// TODO: Later move these functions and structs to dtmSettings.hlsl and a namespace like dtmSettings::height_shading or dtmSettings::contours, etc..

```

Comment:
Please move all the dtm related, fragment shader functions, including the SDFs to another file called `dtm.hlsl` in `namespace dtm`
I don't like how cluttered the fragment_shader is becoming.